reactor-netty-core/src/main/java/reactor/netty/transport/ClientTransport.java

```diff
@@ -140,6 +142,48 @@ public T doOnDisconnected(Consumer<? super Connection> doOnDisconnected) {
 		return dup;
 	}
 
+	/**
+	 * Set or add a callback called before {@link SocketAddress} is resolved.
+	 *
+	 * @param doOnResolve a consumer observing resolve events
+	 * @return a new {@link ClientTransport} reference
+	 */
+	public T doOnResolve(Consumer<? super Channel> doOnResolve) {
```

Comment:
In the `do*` API we work with `Connection` and not `Channel`.
Is it possible to make the solution based on `ConnectionObserver` events?

Reply:
@violetagg I have few questions about this request:

1. Inside `doResolveAndConnect` I do not have access to `Connection`.  Do you have a suggestion on how to achieve it?

2. I would like to be able to pass information from `doOnResolve` to `doAfterResponse`.  The use case is that we would like to time the resolution time and add to the trace.  Unless I put a mutable object inside the Reactor `Context`, I am not sure how I can do it by just relying on `ConnectionObserver`.

3. I see the following logic in `doResolveAndConnect`:

```java
			if (!resolver.isSupported(remoteAddress) || resolver.isResolved(remoteAddress)) {
				MonoChannelPromise monoChannelPromise = new MonoChannelPromise(channel);
				doConnect(remoteAddress, bindAddress, monoChannelPromise);
				return monoChannelPromise;
			}
```

Does this mean that it would call the `resolve` method only once?  The subsequent resolutions would be from cache?  If so, is there an invalidation time for the cache?

Reply:
> @violetagg I have few questions about this request:
> 
> 1. Inside `doResolveAndConnect` I do not have access to `Connection`.  Do you have a suggestion on how to achieve it?

`reactor.netty.Connection#from(io.netty.channel.Channel)`

> 2. I would like to be able to pass information from `doOnResolve` to `doAfterResponse`.  The use case is that we would like to time the resolution time and add to the trace.  Unless I put a mutable object inside the Reactor `Context`, I am not sure how I can do it by just relying on `ConnectionObserver`.

You can either use Reactor Context or Netty Channel Attributes
With channel attributes it is something like `channel.attr(...).get()`

> 3. I see the following logic in `doResolveAndConnect`:
> 
> ```java
> 			if (!resolver.isSupported(remoteAddress) || resolver.isResolved(remoteAddress)) {
> 				MonoChannelPromise monoChannelPromise = new MonoChannelPromise(channel);
> 				doConnect(remoteAddress, bindAddress, monoChannelPromise);
> 				return monoChannelPromise;
> 			}
> ```
> 
> Does this mean that it would call the `resolve` method only once? The subsequent resolutions would be from cache? If so, is there an invalidation time for the cache?

Did you check the new functionality around DNS resolver?
`reactor.netty.transport.NameResolverProvider`
`reactor.netty.transport.ClientTransport#resolver(java.util.function.Consumer<reactor.netty.transport.NameResolverProvider.NameResolverSpec>)`

Reply:
1. Thanks @violetagg!  I just pushed a commit with updates to use `Connection` instead.  I was not aware of the `Connection#from(io.netty.channel.Channel)` API before.

2. I previously misunderstood and thought I would pass `ConnectionObserver` in the lamdbda, that' why couldn't find a way to access to `Channel`.  Now allowing access to `Connection` object, I can use Netty Channel Attributes.

3. Thanks for the pointer.